homeassistant/util/yaml.py

```diff
@@ -140,40 +147,46 @@ def _env_var_yaml(loader: SafeLineLoader,
+    secrets = {}
+    if os.path.isfile(os.path.join(secret_path, _SECRET_YAML)):
+        secrets = load_yaml(
+            os.path.join(secret_path, _SECRET_YAML))
+        if 'logger' in secrets:
+            logger = str(secrets['logger']).lower()
+            if logger == 'debug':
+                _LOGGER.setLevel(logging.DEBUG)
+            else:
+                _LOGGER.error("secrets.yaml: 'logger: debug' expected,"
+                              " but 'logger: %s' found", logger)
+            del secrets['logger']
+    return secrets
+
+
 # pylint: disable=protected-access
 def _secret_yaml(loader: SafeLineLoader,
                  node: yaml.nodes.Node):
     """Load secrets and embed it into the configuration YAML."""
-    # Create secret cache on loader and load secrets.yaml
-    if not hasattr(loader, '_SECRET_CACHE'):
-        loader._SECRET_CACHE = {}
-
-    secret_path = os.path.join(os.path.dirname(loader.name), _SECRET_YAML)
-    if secret_path not in loader._SECRET_CACHE:
-        if os.path.isfile(secret_path):
-            loader._SECRET_CACHE[secret_path] = load_yaml(secret_path)
-            secrets = loader._SECRET_CACHE[secret_path]
-            if 'logger' in secrets:
-                logger = str(secrets['logger']).lower()
-                if logger == 'debug':
-                    _LOGGER.setLevel(logging.DEBUG)
-                else:
-                    _LOGGER.error("secrets.yaml: 'logger: debug' expected,"
-                                  " but 'logger: %s' found", logger)
-                del secrets['logger']
-        else:
-            loader._SECRET_CACHE[secret_path] = None
-    secrets = loader._SECRET_CACHE[secret_path]
-
-    # Retrieve secret, first from secrets.yaml, then from keyring
-    if secrets is not None and node.value in secrets:
-        _LOGGER.debug('Secret %s retrieved from secrets.yaml.', node.value)
-        return secrets[node.value]
-    for sname, sdict in loader._SECRET_CACHE.items():
-        if node.value in sdict:
-            _LOGGER.debug('Secret %s retrieved from secrets.yaml in other '
-                          'folder %s', node.value, sname)
-            return sdict[node.value]
+    secret_path = os.path.dirname(loader.name)
+    while os.path.exists(secret_path) and not secret_path == os.path.dirname(
+            sys.path[0]):
+        _LOGGER.debug('Checking path %s', secret_path)
+        if secret_path not in __SECRET_CACHE:
+            __SECRET_CACHE[secret_path] = _load_secret_yaml(secret_path)
+        secrets = __SECRET_CACHE[secret_path]
+        if node.value in secrets:
+            _LOGGER.debug('Secret %s retrieved from secrets.yaml in '
+                          'folder %s', node.value, secret_path)
+            return secrets[node.value]
+        next_path = os.path.dirname(secret_path)
+
+        if not next_path or next_path == secret_path:
+            # Somehow we got past the .homeassistant configuration folder...
```

Comment:
This probably happens in the unit test case, since in bootstrap we add that folder, but otherwise it will not always be that way

In your unit test prep module you can call bootstrap.mount_local_lib_path(and the path) if you want to try solve it, otherwise I think its good just to break out like you do now